### DISCUSSION TITLE
Access Control for Salvo

### DISCUSSION THREAD

**Discussion by greenpdx** [salvo-rs/salvo#317] [General]
I have different sections of a web app that require different access for different roles.

I would like to see a module that can define access to different section of a web app, The rest/rpc areas validates session data to determine access to a call.   

I am using salvo because it is embedded in a mqtt server and to reduce space and have the web pages effect the mqtt server. use a separate web sever
actix-web

**Additional context**

**Comment by chrislearn** [comment]
I'm not sure how other web frameworks do it, but in Salvo, you can use middleware to define the permissions for each route and its sub-routes.

**Comment by greenpdx** [comment]
Is that middelware available? where is the code?

**Reply by chrislearn** [reply]
I am not entirely clear about your requirements, but I believe that your requirements are specific to your business and not a general requirement. Therefore, you should implement the corresponding middleware yourself based on your business needs.

**Reply by greenpdx** [reply]
You are say It can be done but it has not been done. 
You have login and sessions, but after login every page is accessible to everybody.
This is the security issue of my request.  A general access control middelware that limits access depending upon the session data, 
The configuration can be done in the config.toml file. 
    (path, method, session data[role]) tuple in the config file would define the ACL
On login a role is assigned to that user and stored in the session data
I have not stated any business specific use cases. 
To me this is a general requirement for a web server.

**Reply by chrislearn** [reply]
Salvo already provides commonly used middlewares such as `session`, `basic_auth`, and `jwt_auth`. 

However, specific requirements such as how to log in, configure roles, whether to store data in configuration files or databases, etc., belong to your specific needs, and you can implement them by combining these middlewares yourself. 

Salvo can only provide middleware for general functionality.

**Reply by greenpdx** [reply]
I think there is miscommunication, Yes I am using session and auth, But the next logical step is to provide a general access control.  Login provides two roles, unauthorized and authorized. 

For every use case that requires finer control to access a page is the middleware I am asking for. 

Please tell me why you think only two roles is enough for all use cases. 

Every other web servers provides Access Control.
This will be the last request for this feature, If I did not convince you please close the discussion.

**Reply by josejachuf** [reply]
In another frameworks that I use has the authentication and authorization part included. The structure in DB is similar to this:

auth_users (users)
auth_groups (roles)
auth_permissions (user-role)
and another tables

then, with middleware, you can validate roles (the roles of the logued user)

**Reply by greenpdx** [reply]
I want to keep it simple so there is only roles( groups), if a user needs different access they are given a separate role.  The user has a vec/hash/set of roles that can be used for access.
It also makes it more general.  
a uint16 could define a role giving 64K different possible roles.
to make it simpler a uint64 bitmap could be used.

**Reply by josejachuf** [reply]
something like this should work:

//middleware
#[handler]
pub async fn validate_role(req: &mut Request, depot: &mut Depot, res: &mut Response) {
    let user_key = req.query::<String>("key").unwrap_or("".to_string());

    let role = depot.obtain::<role>().unwrap();
    let user = depot.obtain::<user>().unwrap();

    // pseudocode     
    if role not in user.roles {
        return forbidden
    }
}

in router:
Router::with_path("foo")
    .hoop(affix::insert("role", "admin"))
    .hoop(affix::insert("user", user))
    .hoop(auth::validate_role)
    .push(Router::with_path("<bar>").get(bar::get_all))

**Reply by chrislearn** [reply]
1. I never said "two roles is enough for all use cases.".
2. Please give other frameworks example,  Maybe I don't understand what kind of functionality you want.

**Reply by greenpdx** [reply]
That good.  You are using role as access control, but it hard coded //middleware
I would like dynamic configuration access control.
I think you are better at rust than I am. 
in the config file define the acl
    path, method, role, action (forbid, allow, redirect, function)
I have two function read config to depot.access_role 

Then in validate_role reads depot.access_role and req.path and user from session and does the action.


#[handler]
pub async fn validate_role(req: &mut Request, depot: &mut Depot, res: &mut Response) {
let user_key = req.query::("key").unwrap_or("".to_string());
let access_role = path2access(req, depot) // translate path to role or 

let user = depot.obtain::<user>().unwrap();
// pseudocode     
if access_role not in user.roles {
    return forbidden
}

#[handler]  // read config and setup access control for each path
pub async fn access_cfg(cfg, depot) {
if not depot.access_role {
    read config
   depot.access_role from config
  }
}

in router:
Router::new()
.hoop(auth::access_cfg)
.hoop(auth::validate_role)
.push(Router::with_path("foo").get(bar::get_all))

**Reply by prabirshrestha** [reply]
You can try integrating https://github.com/casbin/casbin-rs or https://github.com/cedar-policy/cedar

**Reply by greenpdx** [reply]
I will look at them
